third_party/rust/adblock/v0_8/crate/src/resources/resource_storage.rs

```diff
@@ -17,6 +17,87 @@ pub struct ResourceStorage {
+    const BS: u8 = b'\\';
+    const BB: u8 = b'b';
+    const TT: u8 = b't';
+    const NN: u8 = b'n';
+    const FF: u8 = b'f';
+    const RR: u8 = b'r';
+    const UU: u8 = b'u';
+    const __: u8 = 0;
+
+    // Look up table for characters that need escaping in a product string
+    static ESCAPED: [u8; 256] = [
+    // 0   1   2   3   4   5   6   7   8   9   A   B   C   D   E   F
+      UU, UU, UU, UU, UU, UU, UU, UU, BB, TT, NN, UU, FF, RR, UU, UU, // 0
+      UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, UU, // 1
+      __, __, QU, __, __, __, __, __, __, __, __, __, __, __, __, __, // 2
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 3
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 4
+      __, __, __, __, __, __, __, __, __, __, __, __, BS, __, __, __, // 5
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 6
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 7
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 8
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // 9
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // A
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // B
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // C
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // D
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // E
+      __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, // F
+    ];
+
+    #[inline(never)]
+    fn write_string_complex(output: &mut Vec<u8>, string: &str, mut start: usize) {
+        output.extend_from_slice(&string.as_bytes()[ .. start]);
+
+        for (index, ch) in string.bytes().enumerate().skip(start) {
+            let escape = ESCAPED[ch as usize];
+            if escape > 0 {
+                output.extend_from_slice(&string.as_bytes()[start .. index]);
+                output.extend_from_slice(&[b'\\', escape]);
+                start = index + 1;
+            }
+            if escape == b'u' {
+                output.extend_from_slice(format!("{:04x}", ch).as_bytes());
+            }
+        }
+        output.extend_from_slice(&string.as_bytes()[start ..]);
+    }
+
+    let mut output = Vec::with_capacity(arg.as_bytes().len() + 2);
+    if QUOTED {
+        output.push(b'"');
+    }
+
+    'process: {
+        for (index, ch) in arg.bytes().enumerate() {
+            if ESCAPED[ch as usize] > 0 {
+                write_string_complex(&mut output, arg, index);
+                break 'process;
+            }
+        }
+
+        output.extend_from_slice(arg.as_bytes());
+    }
+
+    if QUOTED {
+        output.push(b'"');
+    }
+
+    // unwrap safety: input is always valid UTF8; output processing only replaces some ASCII
+    // characters with other valid ones
+    return String::from_utf8(output).unwrap();
```

Comment:
This hurts my brain a bit. The lookup table works on utf-8 bytes because any multi-byte sequences will all have the high bit set, and the table *has no non-zero entries* for bytes with the high bit set? Is that correct?

Reply:
yep, that's exactly right!

Reply:
Ok, thanks. Might be worth expanding the comment a bit, since that invariant is not enforced by code. And/or write a unit test for it.